core/src/main/java/com/gentics/mesh/core/migration/AbstractMigrationHandler.java

```diff
@@ -112,26 +116,35 @@ protected void prepareMigration(HibFieldSchemaVersionElement<?, ?, ?, ?, ?> from
 	 */
 	protected void migrate(NodeMigrationActionContext ac, HibFieldContainer newContainer, FieldContainer newContent,
 						   HibFieldSchemaVersionElement<?, ?, ?, ?, ?> fromVersion) throws Exception {
-		ArrayList<HibFieldSchemaVersionElement<?, ?, ?, ?, ?>> versionChain = new ArrayList<>(1);
+		LinkedList<HibFieldSchemaVersionElement<?, ?, ?, ?, ?>> versionChain = new LinkedList<>();
 		do {
 			versionChain.add(fromVersion);
 			fromVersion = fromVersion.getNextVersion();
 		} while (fromVersion != null && !newContainer.getSchemaContainerVersion().getVersion().equals(fromVersion.getVersion()));
 
+		AtomicReference<FieldContainer> currentContent = new AtomicReference<FieldContainer>(newContent);
+
 		versionChain.stream()
 			.flatMap(version -> version.getChanges().map(change -> Pair.of(change, version)))
 			.filter(pair -> !(pair.getKey() instanceof HibRemoveFieldChange)) // nothing to do for removed fields, they were never added
-			.map(pair -> Pair.of(pair.getValue(), pair.getKey().createFields(pair.getValue().getSchema(), newContent)))
 			.forEach(pair -> {
+				Map<String, Field> currentChanges = pair.getKey().createFields(pair.getValue().getSchema(), currentContent.get());
 				FieldMap fm = new FieldMapImpl();
-				fm.putAll(pair.getValue());
-				
+				fm.putAll(currentChanges);
+
+				// Here we update our content change storage to not loose the changes from the current version
+				currentContent.updateAndGet(oldContent -> () -> oldContent.getFields().putAll(currentChanges));
+
 				// If a migration has been started over non-adjacent from/to versions, the intermediate changes need no actual storage, but a validation..
-				FieldSchemaContainerVersion schema = pair.getKey().getNextVersion().getSchema();
+				FieldSchemaContainerVersion schema = pair.getValue().getNextVersion().getSchema();
 				if (schema instanceof FieldSchemaContainerVersion && !((FieldSchemaContainerVersion)schema).getVersion().equals(newContainer.getSchemaContainerVersion().getVersion())) {
 					log.info("Update skipped for container [{}] of schema [{}] version [{}] from the intermediate version [{}]", newContainer.getUuid(), newContainer.getSchemaContainerVersion().getSchema().getName(), newContainer.getSchemaContainerVersion().getVersion(), ((FieldSchemaContainerVersion)schema).getVersion());
 					schema.assertForUnhandledFields(fm);
 				} else {
+					if (versionChain.size() > 1) {
```

Comment:
No need to iterate over every data change, if there is only one version to upgrade.